flowpm/tfbackground.py

```diff
@@ -0,0 +1,694 @@
+
+    Notes
+    -----
+
+    The expression for :math:`Gf(a)` is:
+
+    .. math::
+        Gf(a)=D'_{1norm}*a**3*E(a)
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    return D1f(cosmo,a)*a**3*E(cosmo,a)
+
+def Gf2(cosmo, a):
+    """
+    FastPM second order growth factor function
+
+    Parameters
+    ----------
+    a : tf.TensorArray
+       Scale factor.
+
+    Returns
+    -------
+    Scalar float Tensor : FastPM second order growth factor function.
+
+    Notes
+    -----
+
+         The expression for :math:`Gf_2(a)` is:
+
+    .. math::
+        Gf_2(a)=D'_{2norm}*a**3*E(a)
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    return D2f(cosmo,a)*a**3*E(cosmo,a)
+
+################################################################
+
+def gf(cosmo, a):
+    """
+    Derivative of Gf against a
+
+            Parameters
+            ----------
+            a : tf.TensorArray
+               Scale factor.
+
+            Returns
+            -------
+            Scalar float Tensor : the derivative of Gf against a.
+
+            Notes
+            -----
+
+         The expression for :math:`gf(a)` is:
+
+    .. math::
+        gf(a)=\frac{dGF}{da}= D^{''}_1 * a ** 3 *E(a) +D'_{1norm}*a ** 3 * E'(a)
+                +   3 * a ** 2 * E(a)*D'_{1norm}
+
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    # Maybe compute ODE or use stored result
+    cache = maybe_compute_ODE(cosmo)
+    lna = tf.math.log(a)
+    d1f = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['D1f'])
+    f1p = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['F1p'])
+    return  (f1p * a ** 3 *E(cosmo,a) +  d1f*a ** 3 * dEa(cosmo,a)  +   3 * a ** 2 * E(cosmo,a)*d1f)
+
+def gf2(cosmo, a):
+    """
+    Derivative of Gf2 against a
+
+            Parameters
+            ----------
+            a : tf.TensorArray
+               Scale factor.
+
+            Returns
+            -------
+            Scalar float Tensor : the derivative of Gf2 against a.
+
+            Notes
+            -----
+
+         The expression for :math:`gf2(a)` is:
+
+    .. math::
+        gf_2(a)=\frac{dGF_2}{da}= D^{''}_2 * a ** 3 *E(a) +D'_{2norm}*a ** 3 * E'(a)
+                +   3 * a ** 2 * E(a)*D'_{2norm}
+
+    """
+    a=tf.convert_to_tensor(a,dtype=tf.float32)
+    # Maybe compute ODE or use stored result
+    cache = maybe_compute_ODE(cosmo)
+    lna = tf.math.log(a)
+    d2f = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['D2f'])
+    f2p = tfp.math.interp_regular_1d_grid(lna,
+                      tf.math.log(cache['a'][0]), tf.math.log(cache['a'][-1]), cache['F2p'])
+    return  (f2p * a ** 3 *E(cosmo,a) +  d2f*a ** 3 * dEa(cosmo,a) + 3 * a ** 2 * E(cosmo,a)*d2f)
+
+#
+# if __name__ == "__main__":
```

Comment:
We can keep this main section and use it to plot a few things for tests if that's useful. If not useful, we sshould remove all of this commented section